```{R}
rm(list=ls())
library(ChIPseeker)
library(ChIPpeakAnno)
library(clusterProfiler)
require(GenomicFeatures)
library(GenomicRanges)
library(readxl)
library(ggplot2)
library(devEMF)
library(Biostrings)
library(memes)
library(dplyr)
getwd()
```




##### Kleb-DAPseq #####
```{r}
# load annotation
hvkp4_txdb <- makeTxDbFromGFF("/Users/lubeifang/Desktop/BIOTOOLS/ref/CHIPPEAKANNO/hvkp4.gff")
hvkp4_annoData <- toGRanges(hvkp4_txdb, feature="gene")
gene_anno <- read.csv("/Users/lubeifang/Desktop/Kleb_dap/ref/hvkp4_geneanno0531.csv")
TF_name <- read.csv("/Users/lubeifang/Desktop/Kleb_dap/files/klebTF_pfam.csv")

# read fasta
fasta_file <- "/Users/lubeifang/Desktop/BIOTOOLS/ref/hvkp4_meme.fasta"
kp_stringset <- readDNAStringSet(fasta_file)


# define peak anno fuction
# For chromosome genes
peak.anno <- function(peak.file){
  peak <- peak.file
  peak_gr <- GRanges(seqnames="chr1",
                   ranges=IRanges(start=as.numeric(as.matrix(peak[,2])),
                                  end= as.numeric(as.matrix(peak[,3])),
                                  names=as.matrix(as.matrix(peak[,4])),
                                  weight=as.numeric(peak[,5])))
  annotatedpeak <- annotatePeakInBatch(peak_gr, AnnotationData=hvkp4_annoData)
  df <- as.data.frame(annotatedpeak)
  return(df)
}

peak.anno_plasmid <- function(peak.file){
  peak <- peak.file
  peak_gr <- GRanges(seqnames=peak.file$chr,
                   ranges=IRanges(start=as.numeric(as.matrix(peak[,2])),
                                  end= as.numeric(as.matrix(peak[,3])),
                                  names=as.matrix(as.matrix(peak[,4])),
                                  weight=as.numeric(peak[,5])))
  annotatedpeak <- annotatePeakInBatch(peak_gr, AnnotationData=hvkp4_annoData)
  df <- as.data.frame(annotatedpeak)
  return(df)
}

```



# Read our peaks
# annotate plasmid gene
```{R}
# read peak
peak.list <- list.files("/Users/lubeifang/Desktop/Kleb_dap/peaks/peak_bed/")
#feature <- read.csv("/Users/lubeifang/Desktop/Kleb_dap/files/feature.csv")
#target <- read.csv("/Users/lubeifang/Desktop/Kleb_dap/files/target.csv")
feature <- data.frame()
target <- data.frame()
for (i in peak.list){
  print(i)
  peaks <- read.csv(paste0("/Users/lubeifang/Desktop/Kleb_dap/peaks/peak_bed/",i),sep = "\t", header = F)
  col_names <- c("chr", "start", "end", "peak_name", "score",
                 "strand", "signalValue", "-log10pvalue", "-log10qvalue", 
                 "relativesummitpositiontopeakstart")
  colnames(peaks) <- col_names
  # exclude the chr
  peaks <- peaks[peaks$`-log10pvalue` >= 2,]
  peaks <- peaks[peaks$chr != "NZ_CP040539.1",]
  if (nrow(peaks) > 0){
    
      peaks$chr <- case_when(
        peaks$chr == "NZ_CP040540.1" ~ "chr2",
        peaks$chr == "NZ_CP040541.1" ~ "chr3",
        peaks$chr == "NZ_CP040542.1" ~ "chr4",
        peaks$chr == "NZ_CP040543.1" ~ "chr5",
        peaks$chr == "NZ_CP040544.1" ~ "chr5",
        TRUE ~ NA_character_ )
      
      peaks.annotated <- peak.anno_plasmid(peaks)
      anno <- merge(peaks,peaks.annotated, by="start", sort = F, all.x = T)
      anno <- merge(anno, gene_anno, by.x = "feature", by.y = "Locus.tag",all.x=T)
      anno <- anno[order(anno$"signalValue",decreasing=T),]
      anno <- anno[,c(3,2,4,1,5:27)]
      name <- sub("\\..*", "", i)
      name <- sub("_.*","",name)
      
      # Change to the gene name
      gene_name <- TF_name[TF_name$label == name, "Gene"]
      anno <- data.frame(TF = gene_name, as.data.frame(anno))
      
      write.csv(anno,paste0("/Users/lubeifang/Desktop/Kleb_dap/peaks/annotated_peak_plasmid/",gene_name,".csv"),row.names = F)
  }
}
```

# annotate chromosome gene
```{R}
# read peak
peak.list <- list.files("/Users/lubeifang/Desktop/Kleb_dap/peaks/peak_bed/")
#feature <- read.csv("/Users/lubeifang/Desktop/Kleb_dap/files/feature.csv")
#target <- read.csv("/Users/lubeifang/Desktop/Kleb_dap/files/target.csv")
feature <- data.frame()
target <- data.frame()
for (i in peak.list){
  print(i)
  peaks <- read.csv(paste0("/Users/lubeifang/Desktop/Kleb_dap/peaks/peak_bed/",i),sep = "\t", header = F)
  col_names <- c("chr", "start", "end", "peak_name", "score",
                 "strand", "signalValue", "-log10pvalue", "-log10qvalue", 
                 "relativesummitpositiontopeakstart")
  colnames(peaks) <- col_names
  # Only keep the chr
  peaks <- subset(peaks,chr =="NZ_CP040539.1")
  peaks <- peaks[peaks$`-log10pvalue` >= 2,]
  peaks.annotated <- peak.anno(peaks)
  anno <- merge(peaks,peaks.annotated, by="start", sort = F, all.x = T)
  anno <- merge(anno, gene_anno, by.x = "feature", by.y = "Locus.tag",all.x=T)
  anno <- anno[order(anno$"signalValue",decreasing=T),]
  anno <- anno[,c(3,2,4,1,5:27)]
  name <- sub("\\..*", "", i)
  name <- sub("_.*","",name)
  
  # Change to the gene name
  gene_name <- TF_name[TF_name$label == name, "Gene"]
  anno <- data.frame(TF = gene_name, as.data.frame(anno))
  
#  write.csv(anno,paste0("/Users/lubeifang/Desktop/Kleb_dap/peaks/annotated_peak/",gene_name,".csv"),row.names = F)

# write fasta
  gR <- GRanges(seqnames = "chr1",
                ranges = IRanges(start = anno$start, end = anno$end.x),
                strand = "*")
  gR <- resize(gR,50, "center")
  sequences <- get_sequence(gR, kp_stringset)
  writeXStringSet(sequences, paste0("/Users/lubeifang/Desktop/Kleb_dap/peaks/peak_fasta/",gene_name,".fasta"), append=FALSE,
                compress=FALSE, compression_level=NA, format="fasta")

  # select promoter region target
  anno_target <- subset(anno, insideFeature == "overlapStart" | insideFeature == "upstream", 
                        select = c("feature", "signalValue", "Symbol"))
  a=rep(gene_name, times = nrow(anno_target))
  anno_target <- cbind(data.frame(TF_candidate = a), anno_target)
  target <- rbind(target, anno_target)
  frequency_table <- table(as.data.frame(anno)$insideFeature)
  # 将频数转换为占比
  percentage_table <- prop.table(frequency_table)
  feature <- rbind(feature, data.frame(Category = gene_name, Percentage = percentage_table))
}


# write.csv(feature,"/Users/lubeifang/Desktop/Kleb_dap/files/feature.csv", row.names = FALSE)
# write.csv(target,"/Users/lubeifang/Desktop/Kleb_dap/files/target.csv", row.names = FALSE, quote = F)
```